### ISSUE TITLE
Use size_t instead of unsigned int and int in gldcore/load.c

### ISSUE THREAD

**Issue by esilksel** [gridlab-d/gridlab-d#1225]
Initial Checklist prior to submitting the issue. Check the bellow boxes to the best of your knowledge:

- [x] I have not found an open issue describing the same problem.
- [x] I went through "CONTRIBUTING.MD" before deciding to submit the issue.
- [x] While building GridLAB-D, I followed all steps as detailed in GridLAB-D Wiki.

------------------------------------------------------------------------------------------------------------

## Please describe what was the cause of the issue?
In investigating another issue (to be submitted at a later date, still waiting on more background from a user), it was discovered that `load.c` is using `unsigned int` for the `object_index_size`, a variable used to specify the size of `OBJECT **object_index` during `malloc`/`realloc` calls. Additionally, the `new_size` variable is of type `int`.

On a 64 bit machine, this can cause the the program to declare it is out of memory well before that is actually the case.

## Please provide the steps to reproduce.
The easiest method to reproduce this is to provide a `.glm` file with large ID present. In particular, our test file is complaining about the following entry:
```
object meter:2010000000 { name TS01DS00; phases ABCN; nominal_voltage 79600; }
```

## Please provide the exact error/warnining/failure message printed on the console.
```
Gen100kLines.glm(56654): unable to index object id number for meter:2010000000
Gen100kLines.glm(56654): load failed at or near 'object meter...'
ERROR    [INIT] : unable to load 'Gen100kLines.glm': Cannot allocate memory
FATAL    [INIT] : shutdown after command line rejected
```

## If possible, please provide any screenshots and other supporting information.
Making the change to `size_t` when dealing with arrays and calls to `sizeof()` appears to have corrected this, and the model is capable of running (although the model in question is failing to converge).

The specific function (with the proposed changes):
```
static OBJECT **object_index = NULL;
static unsigned char *object_linked = NULL;
static size_t object_index_size = 65536;  // CHANGE HERE
/*static*/ STATUS load_set_index(OBJECT *obj, OBJECTNUM id)
{
    if (object_index==NULL)
    {
        object_index = malloc(sizeof(OBJECT*)*object_index_size);
        memset(object_index,0,sizeof(OBJECT*)*object_index_size);
        object_linked = malloc(sizeof(unsigned char)*object_index_size);
        memset(object_linked,0,sizeof(unsigned char)*object_index_size);
    }
    if (id>=object_index_size) /* index needs to grow */
    {
        //int new_size = (id/object_index_size+1)*object_index_size;
        size_t new_size = (id/object_index_size+1)*object_index_size;
        fprintf(stderr, "new_size: %llu\n", new_size);
        object_index = realloc(object_index,sizeof(OBJECT*)*new_size);
        object_linked = realloc(object_linked,sizeof(unsigned char)*new_size);
        object_index_size = new_size;
    }
    if (object_index==NULL) { errno = ENOMEM; return FAILED;}
    /* collision check here */
    object_index[id] = obj;
    object_linked[id] = 0;
    return SUCCESS;
}
```
It may be beneficial to review the code base and determine if there are other locations where `size_t` would be the correct type.

-------------------------------------------------------------------------------------------------------------

**Comment by ftuffner** [comment]
@eberleim Is this related to the similar issue you just fixed recently (and may still be in a pull request)?

**Comment by ftuffner** [comment]
@esilksel A quick "work-around", and actually preferred method for using GridLAB-D objects, is to not specify the object number.  GridLAB-D will assign that automatically internally.  For any object references, (like a `from` or `to` on a line), just use the object name.  I know not all of our files are updated to reflect this.

As a simple example:
```
object node:1 {
	phases "ABCN";
	nominal_voltage 7200;
}

object overhead_line:12 {
	phases "ABCN";
	from node:1;
	to node:2;
	length 2000;
	configuration line_configuration:300;
}

object node:2 {
	phases "ABCN";
	nominal_voltage 7200;
}
```

would be:

```
object node {
	name node_1;
	phases "ABCN";
	nominal_voltage 7200;
}

object overhead_line {
	name ohl_12;
	phases "ABCN";
	from node_1;
	to node_2;
	length 2000;
	configuration line_config_300;
}

object node {
	name node_2;
	phases "ABCN";
	nominal_voltage 7200;
}
```

We'll still look into the above issue, but removing the explicit object ID references helps eliminate "exciting" it.

**Comment by esilksel** [comment]
While we are doing some very large models, we're not hitting the 2^32 limit imposed by the choice of variable type on our machines. So, I'll certainly defer to you as to whether or not this _needs_ to be implemented.

As for the suggested fix: yes, this will solve the immediate issue. However, our power engineers have expressed a need to use the object ID's as ID's, rather than just a count of objects. I'll collect some more information from them if needed, but the short version is that they're using them like serial numbers (with types, branch identifiers, etc.) in order to aid in system generation and faster debugging. We can probably solve this locally by assigning the ID as a count and using it as an index into a table of the preferred format ID's (e.g. object 5 -> object 500202401, etc.) that is generated alongside. But, their intuition told them an object ID was separable from the object count.

**Comment by esilksel** [comment]
Got an example of why they are interested in hand-naming the objects. When the forward-backward solver fails to converge, it complains like so:
```
ERROR    [timestamp] : convergence iteration limit reached for object meter:3045
ERROR    [timestamp] : last error message was repeated NNNN times
```
In this case, the use is confident it is a topology problem...but are unable to determine which object(s) are actually to blame.

**Comment by ftuffner** [comment]
@esilksel That issue should be resolvable in "better error messages".  We try to make it so all of them dump the object ID and name, but I know there are some legacy messages that don't have that capability.  Part of this issue looks to be a "double-check the various error messages for completeness" on our end.

**Comment by eberleim** [comment]
I've looked into this, and while technically changing to `size_t` (alias for `unsigned long long`) does resolve the visible issue here, the whole behavior is dangerous at best. 
The code does two `malloc/realloc(sizeof(void*) * object_size_index)` equivalent operations, so with your example value of 2010000000, each of those malloc/realloc will allocate 2010000000*8 bytes, or approx 16GB of memory on the heap. 
I would suggest against long indexes at this time, and would suggest a new improvement be requested here to resolve the unbounded memory allocation.

**Comment by esilksel** [comment]
@ftuffner Our users here think that's an agreeable solution. We just need to be able to determine which objects are problematic in a given simulation.
@eberleim Yes, I think ideally it would be preferable to determine the total number of objects separately from their ID's and allocate based upon that. Do you want me to open that enhancement, or is that something that needs to be done internally?
